File: mobile-sdk/ios/ThronosSDK.swift
/**
 * Thronos SDK for iOS
 * Native Swift implementation for iOS applications
 */

import Foundation
import Security

public class ThronosSDK {

    // MARK: - Properties

    private let apiUrl: String
    private let network: String
    private var currentWallet: Wallet?

    // MARK: - Initialization

    public init(apiUrl: String = "http://localhost:5000", network: String = "mainnet") {
        self.apiUrl = apiUrl
        self.network = network

        // Load wallet from keychain
        self.currentWallet = loadWalletFromKeychain()
    }

    // MARK: - Wallet Management

    /// Create a new wallet
    public func createWallet(completion: @escaping (Result<Wallet, Error>) -> Void) {
        guard let url = URL(string: "\(apiUrl)/api/wallet/create") else {
            completion(.failure(ThronosError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }

            guard let data = data else {
                completion(.failure(ThronosError.noData))
                return
            }

            do {
                let wallet = try JSONDecoder().decode(Wallet.self, from: data)
                self?.currentWallet = wallet
                self?.saveWalletToKeychain(wallet)
                completion(.success(wallet))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }

    /// Import an existing wallet
    public func importWallet(address: String, secret: String) -> Result<Wallet, Error> {
        guard address.hasPrefix("THR") else {
            return .failure(ThronosError.invalidAddress)
        }

        let wallet = Wallet(address: address, secret: secret)
        self.currentWallet = wallet
        saveWalletToKeychain(wallet)

        return .success(wallet)
    }

    /// Get current wallet
    public func getWallet() -> Wallet? {
        return currentWallet
    }

    /// Check if wallet is connected
    public func isConnected() -> Bool {
        return currentWallet != nil
    }

    /// Disconnect wallet
    public func disconnect() {
        currentWallet = nil
        deleteWalletFromKeychain()
    }

    // MARK: - Token Operations

    /// Get token balances
    public func getBalances(address: String? = nil, showZero: Bool = false, completion: @escaping (Result<TokensResponse, Error>) -> Void) {
        let walletAddress = address ?? currentWallet?.address

        guard let walletAddress = walletAddress else {
            completion(.failure(ThronosError.noWallet))
            return
        }

        guard let url = URL(string: "\(apiUrl)/api/wallet/tokens/\(walletAddress)?show_zero=\(showZero)") else {
            completion(.failure(ThronosError.invalidURL))
            return
        }

        URLSession.shared.dataTask(with: url) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }

            guard let data = data else {
                completion(.failure(ThronosError.noData))
                return
            }

            do {
                let tokensResponse = try JSONDecoder().decode(TokensResponse.self, from: data)
                completion(.success(tokensResponse))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }

    /// Send transaction
    public func sendTransaction(to: String, amount: Double, token: String = "THR", completion: @escaping (Result<TransactionResponse, Error>) -> Void) {
        guard let wallet = currentWallet else {
            completion(.failure(ThronosError.noWallet))
            return
        }

        guard let url = URL(string: "\(apiUrl)/api/wallet/send") else {
            completion(.failure(ThronosError.invalidURL))
            return
        }

        let transaction = TransactionRequest(
            from: wallet.address,
            to: to,
            amount: amount,
            token: token,
            secret: wallet.secret
        )

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        do {
            request.httpBody = try JSONEncoder().encode(transaction)
        } catch {
            completion(.failure(error))
            return
        }

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }

            guard let data = data else {
                completion(.failure(ThronosError.noData))
                return
            }

            do {
                let txResponse = try JSONDecoder().decode(TransactionResponse.self, from: data)
                completion(.success(txResponse))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }

    // MARK: - Keychain Management

    private func saveWalletToKeychain(_ wallet: Wallet) {
        let addressData = wallet.address.data(using: .utf8)!
        let secretData = wallet.secret.data(using: .utf8)!

        // Save address
        let addressQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "thronos_wallet_address",
            kSecValueData as String: addressData
        ]

        SecItemDelete(addressQuery as CFDictionary)
        SecItemAdd(addressQuery as CFDictionary, nil)

        // Save secret
        let secretQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "thronos_wallet_secret",
            kSecValueData as String: secretData
        ]

        SecItemDelete(secretQuery as CFDictionary)
        SecItemAdd(secretQuery as CFDictionary, nil)
    }

    private func loadWalletFromKeychain() -> Wallet? {
        // Load address
        let addressQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "thronos_wallet_address",
            kSecReturnData as String: true
        ]

        var addressResult: AnyObject?
        let addressStatus = SecItemCopyMatching(addressQuery as CFDictionary, &addressResult)

        guard addressStatus == errSecSuccess,
              let addressData = addressResult as? Data,
              let address = String(data: addressData, encoding: .utf8) else {
            return nil
        }

        // Load secret
        let secretQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "thronos_wallet_secret",
            kSecReturnData as String: true
        ]

        var secretResult: AnyObject?
        let secretStatus = SecItemCopyMatching(secretQuery as CFDictionary, &secretResult)

        guard secretStatus == errSecSuccess,
              let secretData = secretResult as? Data,
              let secret = String(data: secretData, encoding: .utf8) else {
            return nil
        }

        return Wallet(address: address, secret: secret)
    }

    private func deleteWalletFromKeychain() {
        let addressQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "thronos_wallet_address"
        ]
        SecItemDelete(addressQuery as CFDictionary)

        let secretQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "thronos_wallet_secret"
        ]
        SecItemDelete(secretQuery as CFDictionary)
    }
}

// MARK: - Models

public struct Wallet: Codable {
    public let address: String
    public let secret: String
}

public struct Token: Codable {
    public let symbol: String
    public let name: String
    public let balance: Double
    public let decimals: Int
    public let color: String
    public let logo: String?
}

public struct TokensResponse: Codable {
    public let address: String
    public let tokens: [Token]
    public let last_updated: String
}

public struct TransactionRequest: Codable {
    public let from: String
    public let to: String
    public let amount: Double
    public let token: String
    public let secret: String
}

public struct TransactionResponse: Codable {
    public let success: Bool
    public let transaction: Transaction?
}

public struct Transaction: Codable {
    public let hash: String
    public let from: String
    public let to: String
    public let amount: Double
    public let token: String
    public let timestamp: String
}

// MARK: - Errors

public enum ThronosError: Error {
    case invalidURL
    case noData
    case noWallet
    case invalidAddress

    var localizedDescription: String {
        switch self {
        case .invalidURL:
            return "Invalid API URL"
        case .noData:
            return "No data received from server"
        case .noWallet:
            return "No wallet connected"
        case .invalidAddress:
            return "Invalid wallet address"
        }
    }
}

File: mobile-sdk/android/ThronosSDK.kt
/**
 * Thronos SDK for Android
 * Native Kotlin implementation for Android applications
 */

package com.thronos.sdk

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException

class ThronosSDK(
    private val context: Context,
    private val apiUrl: String = "http://localhost:5000",
    private val network: String = "mainnet"
) {

    private val gson = Gson()
    private val client = OkHttpClient()
    private var currentWallet: Wallet? = null
    private val preferences: SharedPreferences

    init {
        // Initialize encrypted shared preferences
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()

        preferences = EncryptedSharedPreferences.create(
            context,
            "thronos_wallet_prefs",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )

        // Load wallet from storage
        currentWallet = loadWalletFromStorage()
    }

    // MARK: - Wallet Management

    /**
     * Create a new wallet
     */
    suspend fun createWallet(): Result<Wallet> = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url("$apiUrl/api/wallet/create")
                .post("".toRequestBody("application/json".toMediaTypeOrNull()))
                .build()

            val response = client.newCall(request).execute()
            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Failed to create wallet"))
            }

            val wallet = gson.fromJson(response.body?.string(), Wallet::class.java)
            currentWallet = wallet
            saveWalletToStorage(wallet)

            Result.success(wallet)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Import an existing wallet
     */
    fun importWallet(address: String, secret: String): Result<Wallet> {
        if (!address.startsWith("THR")) {
            return Result.failure(Exception("Invalid address format"))
        }

        val wallet = Wallet(address, secret)
        currentWallet = wallet
        saveWalletToStorage(wallet)

        return Result.success(wallet)
    }

    /**
     * Get current wallet
     */
    fun getWallet(): Wallet? = currentWallet

    /**
     * Check if wallet is connected
     */
    fun isConnected(): Boolean = currentWallet != null

    /**
     * Disconnect wallet
     */
    fun disconnect() {
        currentWallet = null
        deleteWalletFromStorage()
    }

    // MARK: - Token Operations

    /**
     * Get token balances
     */
    suspend fun getBalances(
        address: String? = null,
        showZero: Boolean = false
    ): Result<TokensResponse> = withContext(Dispatchers.IO) {
        try {
            val walletAddress = address ?: currentWallet?.address
                ?: return@withContext Result.failure(Exception("No wallet connected"))

            val request = Request.Builder()
                .url("$apiUrl/api/wallet/tokens/$walletAddress?show_zero=$showZero")
                .get()
                .build()

            val response = client.newCall(request).execute()
            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Failed to get balances"))
            }

            val tokensResponse = gson.fromJson(response.body?.string(), TokensResponse::class.java)
            Result.success(tokensResponse)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Get balance for specific token
     */
    suspend fun getTokenBalance(tokenSymbol: String, address: String? = null): Result<Double> {
        val balancesResult = getBalances(address, true)
        return if (balancesResult.isSuccess) {
            val token = balancesResult.getOrNull()?.tokens?.find { it.symbol == tokenSymbol }
            Result.success(token?.balance ?: 0.0)
        } else {
            balancesResult.map { 0.0 }
        }
    }

    /**
     * Send transaction
     */
    suspend fun sendTransaction(
        to: String,
        amount: Double,
        token: String = "THR"
    ): Result<TransactionResponse> = withContext(Dispatchers.IO) {
        try {
            val wallet = currentWallet
                ?: return@withContext Result.failure(Exception("No wallet connected"))

            val transaction = TransactionRequest(
                from = wallet.address,
                to = to,
                amount = amount,
                token = token,
                secret = wallet.secret
            )

            val requestBody = gson.toJson(transaction)
                .toRequestBody("application/json".toMediaTypeOrNull())

            val request = Request.Builder()
                .url("$apiUrl/api/wallet/send")
                .post(requestBody)
                .build()

            val response = client.newCall(request).execute()
            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Transaction failed"))
            }

            val txResponse = gson.fromJson(response.body?.string(), TransactionResponse::class.java)
            Result.success(txResponse)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Get transaction history
     */
    suspend fun getTransactionHistory(
        address: String? = null,
        limit: Int = 50
    ): Result<List<Transaction>> = withContext(Dispatchers.IO) {
        try {
            val walletAddress = address ?: currentWallet?.address
                ?: return@withContext Result.failure(Exception("No wallet connected"))

            val request = Request.Builder()
                .url("$apiUrl/api/transactions/$walletAddress?limit=$limit")
                .get()
                .build()

            val response = client.newCall(request).execute()
            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Failed to get transaction history"))
            }

            val transactions = gson.fromJson(response.body?.string(), Array<Transaction>::class.java).toList()
            Result.success(transactions)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // MARK: - Storage Management

    private fun saveWalletToStorage(wallet: Wallet) {
        preferences.edit().apply {
            putString("thronos_wallet_address", wallet.address)
            putString("thronos_wallet_secret", wallet.secret)
            apply()
        }
    }

    private fun loadWalletFromStorage(): Wallet? {
        val address = preferences.getString("thronos_wallet_address", null)
        val secret = preferences.getString("thronos_wallet_secret", null)

        return if (address != null && secret != null) {
            Wallet(address, secret)
        } else {
            null
        }
    }

    private fun deleteWalletFromStorage() {
        preferences.edit().apply {
            remove("thronos_wallet_address")
            remove("thronos_wallet_secret")
            apply()
        }
    }

    // MARK: - Models

    data class Wallet(
        val address: String,
        val secret: String
    )

    data class Token(
        val symbol: String,
        val name: String,
        val balance: Double,
        val decimals: Int,
        val color: String,
        val logo: String?
    )

    data class TokensResponse(
        val address: String,
        val tokens: List<Token>,
        @SerializedName("last_updated") val lastUpdated: String
    )

    data class TransactionRequest(
        val from: String,
        val to: String,
        val amount: Double,
        val token: String,
        val secret: String
    )

    data class TransactionResponse(
        val success: Boolean,
        val transaction: Transaction?
    )

    data class Transaction(
        val hash: String,
        val from: String,
        val to: String,
        val amount: Double,
        val token: String,
        val timestamp: String,
        val status: String
    )
}

// MARK: - Extension Functions

/**
 * Extension function to format wallet address
 */
fun String.toShortAddress(prefixLength: Int = 10, suffixLength: Int = 8): String {
    return if (this.length > prefixLength + suffixLength) {
        "${this.substring(0, prefixLength)}...${this.substring(this.length - suffixLength)}"
    } else {
        this
    }
}

/**
 * Extension function to validate Thronos address
 */
fun String.isValidThronosAddress(): Boolean {
    return this.startsWith("THR") && this.length > 10
}
